src/v/json/allocator.h

```diff
@@ -0,0 +1,48 @@
+#include "vlog.h"
+
+#include <fmt/format.h>
+#include <rapidjson/allocators.h>
+namespace json {
+
+class bad_alloc_with_msg final : public std::bad_alloc {
+public:
+    explicit bad_alloc_with_msg(std::string msg)
+      : _msg(std::move(msg)) {}
+
+    const char* what() const noexcept override { return _msg.c_str(); }
+
+private:
+    std::string _msg;
```

Comment:
we don't use `std::string` we use `ss::string`

Reply:
are these strings really dynamic? or are they all defined up front in code. if they are just regular static strings, you can avoid the allocation all together. by returning a static string.

Reply:
`fmt_with_ctx` is usually used with `fmt::format`, which returns a dynamic `std::string`. I guess it should be possible to use `ssx::sformat`, and it looks like there are some usages in serde and scram.

I guess allocating any string during `bad_alloc` is quite possibly not going to work, which is the likely reason `std::bad_alloc` doesn't accept one.

Reply:
perhaps for bad_alloc_with_msg we should store a `const char*` and then use __FILE__ and __LINE__ etc... to build the format string a string literal to avoid the allocations.

Reply:
or maybe that would just give us the file/line of the header. just returning bad_alloc is probably good enough.